lib/wibox/layout/align.lua

```diff
@@ -41,121 +43,297 @@ local align = {}
-                --  widget to take up any space (as the first if block is skipped
-                --  if this is the case)
-                if self._private.expand == "inside" or not self._private.second then
-                    size_remains = size_remains - h
-                end
-            else
-                w, _ = base.fit_widget(self, context, self._private.first, size_remains, height)
-                size_first = w
-                if self._private.expand == "inside" or not self._private.second then
-                    size_remains = size_remains - w
-                end
-            end
+    -- Place a child widget inside the layout
+    local function place_child(child, offset, size, update_size_remains)
+        -- Can't place a widget if it does not exist or there is no more size
+        if not self._private[child] or size_remains <= 0 then
+            return
+        end
+
+        offset = floor(offset)
+        if not size then
+            size = size and floor(size) or size_remains
+            size = fit_child(child, size)
         else
-            if self._private.dir == "y" then
-                h = size_remains
-            else
-                w = size_remains
-            end
+            fit_child(child, size)
+        end
+
+        local child_width = is_vert and width or size
+        local child_height = is_vert and size or height
+
+        local x_offset = (not is_vert) and offset or 0
+        local y_offset = (not is_vert) and 0 or offset
+
+        table.insert(
+            result,
+            base.place_widget_at(
+                self._private[child],
+                x_offset, y_offset,
+                child_width, child_height
+            )
+        )
+
+        if update_size_remains == nil or update_size_remains == true then
+            size_remains = size_remains - size
         end
-        table.insert(result, base.place_widget_at(self._private.first, 0, 0, w, h))
     end
-    -- size_remains will be <= 0 if first used all the space
-    if self._private.third and size_remains > 0 then
-        local w, h, _ = width, height, nil
-        if self._private.expand ~= "outside" then
-            if self._private.dir == "y" then
-                _, h = base.fit_widget(self, context, self._private.third, width, size_remains)
-                -- give the middle widget the rest of the space for "inside" mode
-                if self._private.expand == "inside" then
-                    size_remains = size_remains - h
-                end
-            else
-                w, _ = base.fit_widget(self, context, self._private.third, size_remains, height)
-                if self._private.expand == "inside" then
-                    size_remains = size_remains - w
-                end
-            end
+
+    -- Size of the layout before any placements
+    local size_total = size_remains
+
+    if self._private.expand == "none" then
```

Comment:
That's not your work here, but while we refactor this code: I think I'd like to see each branch of this huge state-machine to be resolved in separates utils functions to make the code easier to read.

(The association of modes and "sub-layout" functions could even be done with some sort of polymorphism :thinking:)

Reply:
Maybe. Another possible solution would be to create a table where the keys are expand modes and the values are their corresponding callback util functions. I imagine this would work well with the enum like table made in [b37ba4e](https://github.com/awesomeWM/awesome/pull/3588/commits/b37ba4e4a54266187492729c2d42691c54ffddc2)

Reply:
Lines like `expand_mode_impl[self._private.expand](...)` are cute, but they increase the complexity of reading/comprehending a function quite a lot.
Aire-One's suggestion is much easier to skim:

```lua
if expand == "none" then
  layout_none(...)
elseif expand == "inside" then
  layout_inside(...)
end
```